src/models/onnxruntime_api.h

```diff
@@ -68,24 +68,114 @@ p_session_->Run(nullptr, input_names, inputs, std::size(inputs), output_names, o
 #include <string>
 #include <vector>
 #include <unordered_map>
+#include <array>
 
 #include "onnxruntime_c_api.h"
 #include "../span.h"
+#include "../logging.h"
 
 #if defined(__ANDROID__)
 #include <android/log.h>
 #include <dlfcn.h>
+
+#define TAG "GenAI"
+
+#define LOG_DEBUG(...) __android_log_print(ANDROID_LOG_DEBUG, TAG, __VA_ARGS__)
+#define LOG_INFO(...) __android_log_print(ANDROID_LOG_INFO, TAG, __VA_ARGS__)
+#define LOG_WARN(...) __android_log_print(ANDROID_LOG_WARN, TAG, __VA_ARGS__)
+#define LOG_ERROR(...) __android_log_print(ANDROID_LOG_ERROR, TAG, __VA_ARGS__)
+#define LOG_FATAL(...) __android_log_print(ANDROID_LOG_FATAL, TAG, __VA_ARGS__)
+
+#elif defined(__linux__)
+#include <dlfcn.h>
+
+#ifndef PATH_MAX
+#define PATH_MAX (4096)
+#endif
+
+#define LOG_DEBUG(...) Generators::Log("debug", __VA_ARGS__)
+#define LOG_INFO(...) Generators::Log("info", __VA_ARGS__)
+#define LOG_WARN(...) Generators::Log("warning", __VA_ARGS__)
+#define LOG_ERROR(...) Generators::Log("error", __VA_ARGS__)
+#define LOG_FATAL(...) Generators::Log("fatal", __VA_ARGS__)
+
 #endif
 
 /** \brief Free functions and a few helpers are defined inside this namespace. Otherwise all types are the C API types
  *
  */
 namespace Ort {
 
+using OrtApiBaseFn = const OrtApiBase* (*)(void);
+
 /// Before using this C++ wrapper API, you MUST call Ort::InitApi to set the below 'api' variable
 inline const OrtApi* api{};
+
+#if defined(__linux__)
+inline void* LoadDynamicLibraryIfExists(const std::string& path) {
+  LOG_INFO("Attempting to dlopen %s native library", path.c_str());
+  void* ort_lib_handle = dlopen(path.c_str(), RTLD_NOW | RTLD_LOCAL);
+  if (ort_lib_handle == nullptr) {
+    return nullptr;
+  }
+
+#if !defined(__ANDROID__)  // RTLD_DI_ORIGIN not available on Android
+  char pathname[PATH_MAX];
+  dlinfo((void*)ort_lib_handle, RTLD_DI_ORIGIN, &pathname);
+  LOG_INFO("Loaded native library at %s", pathname);
+#endif
+  return ort_lib_handle;
+}
+
+inline std::string GetCurrentModuleDir() {
```

Comment:
This is to find the location of the current ort-genai library, in order to locate the `site-pacakges` and `onnxruntime` folder in it.